server/datastore/mysql/software.go

```diff
@@ -517,16 +476,112 @@ func insertNewInstalledHostSoftwareDB(
 		return incomingOrdered[i].uniqueName < incomingOrdered[j].uniqueName
 	})
 
-	for _, s := range incomingOrdered {
+	// Compute checksums for all incoming software, which we will use for faster retrieval, since checksum is a unique index
+	softwareChecksums := make(map[string]*fleet.Software, 0)
+	for i, s := range incomingOrdered {
 		if _, ok := currentMap[s.uniqueName]; !ok {
-			id, err := getOrGenerateSoftwareIdDB(ctx, tx, s.software)
+			checksum, err := computeRawChecksum(s.software)
 			if err != nil {
 				return nil, err
 			}
-			insertsHostSoftware = append(insertsHostSoftware, hostID, id, s.software.LastOpenedAt)
+			softwareChecksums[string(checksum)] = &incomingOrdered[i].software // we can't use `s` here because it is a pointer to a loop variable
+		}
+	}
 
-			s.software.ID = id
-			insertedSoftware = append(insertedSoftware, s.software)
+	// First, we check if host's new software already exists in the software table.
+	if len(softwareChecksums) > 0 {
+		keys := make([]string, 0, len(softwareChecksums))
+		for checksum := range softwareChecksums {
+			keys = append(keys, checksum)
+		}
+		// We use the replica DB for retrieval to minimize the traffic to the master DB.
+		// It is OK if the software is not found in the replica DB, because we will then attempt to insert it in the master DB.
+		existingSoftware, err := getSoftwareIDsByChecksums(ctx, ds.reader(ctx), keys)
+		if err != nil {
+			return nil, err
+		}
+		for _, s := range existingSoftware {
+			software, ok := softwareChecksums[s.Checksum]
+			if !ok {
+				return nil, ctxerr.New(ctx, fmt.Sprintf("software not found for checksum %s", hex.EncodeToString([]byte(s.Checksum))))
```

Comment:
Am a bit confused, why should this return an error? (`softwareChecksums` is incoming software and `existingSoftware` is from the DB)

Reply:
It shouldn't. Just being paranoid.